File: rust/2024/src/bin/day08.rs
struct Frequency {
    data: [(i8, i8); 4],
    len: u8,
}

impl Frequency {
    fn new() -> Self {
        Self {
            data: [(0, 0); 4],
            len: 0,
        }
    }

    fn push(&mut self, point: (i8, i8)) {
        self.data[self.len as usize] = point;
        self.len += 1;
    }

    fn iter(&self) -> std::slice::Iter<'_, (i8, i8)> {
        self.data[..self.len as usize].iter()
    }
}

struct Set {
    data: [u64; 50],
}

impl Set {
    fn new() -> Self {
        Self { data: [0; 50] }
    }

    fn insert(&mut self, point: (i8, i8)) {
        self.data[point.0 as usize] |= 1 << point.1 as usize;
    }

    fn len(&self) -> u32 {
        self.data.iter().map(|s| s.count_ones()).sum()
    }
}

struct Solution {
    frequencies: [Frequency; (b'z' - b'0') as usize],
    size: i8,
}

impl Solution {
    #[inline(always)]
    fn in_range(&self, point: (i8, i8)) -> bool {
        0 <= point.0 && point.0 < self.size && 0 <= point.1 && point.1 < self.size
    }
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        let mut frequencies = std::array::from_fn(|_| Frequency::new());
        let mut x = 0;
        let mut y = 0;

        for b in input.bytes() {
            if b == b'\n' {
                x = 0;
                y += 1;
            } else if b != b'.' {
                frequencies[(b - b'0') as usize].push((x, y));
                x += 1;
            } else {
                x += 1;
            }
        }

        Ok(Self {
            frequencies,
            size: y,
        })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        let mut antinodes = Set::new();

        for frequency in &self.frequencies {
            for ((ax, ay), (bx, by)) in frequency.iter().tuple_combinations() {
                let antinode_a = (2 * ax - bx, 2 * ay - by);
                let antinode_b = (2 * bx - ax, 2 * by - ay);

                if self.in_range(antinode_a) {
                    antinodes.insert(antinode_a);
                }

                if self.in_range(antinode_b) {
                    antinodes.insert(antinode_b);
                }
            }
        }

        Ok(antinodes.len())
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        let mut antinodes = Set::new();

        for frequency in &self.frequencies {
            for ((ax, ay), (bx, by)) in frequency.iter().tuple_combinations() {
                let mut antinode = (*ax, *ay);
                let dx = bx - ax;
                let dy = by - ay;

                while self.in_range(antinode) {
                    antinodes.insert(antinode);
                    antinode.0 -= dx;
                    antinode.1 -= dy;
                }

                antinode = (ax + dx, ay + dy);

                while self.in_range(antinode) {
                    antinodes.insert(antinode);
                    antinode.0 += dx;
                    antinode.1 += dy;
                }
            }
        }

        Ok(antinodes.len())
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = r"............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............
";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "14");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "34");
    }
}

File: python/2024/day08.py
from __future__ import annotations

from collections import defaultdict
from itertools import combinations
from typing import TYPE_CHECKING

from lib import Point

if TYPE_CHECKING:
    from collections.abc import Iterable


class Frequencies:
    freq_groups: defaultdict[str, list[Point]]
    size: int

    def __init__(self, input: str) -> None:
        self.freq_groups = defaultdict(lambda: [])

        for y, line in enumerate(input.splitlines()):
            for x, ch in enumerate(line):
                if ch != ".":
                    self.freq_groups[ch].append(Point(x, y))

        self.size = len(input.splitlines())

    def in_range(self, p: Point) -> bool:
        return 0 <= p.x < self.size and 0 <= p.y < self.size

    def groups(self) -> Iterable[list[Point]]:
        return self.freq_groups.values()


def part1(input: str) -> int:
    frequencies = Frequencies(input)
    antinodes = set()

    for frequency in frequencies.groups():
        for a, b in combinations(frequency, 2):
            antinode_a = a * 2 - b
            antinode_b = b * 2 - a

            if frequencies.in_range(antinode_a):
                antinodes.add(antinode_a)

            if frequencies.in_range(antinode_b):
                antinodes.add(antinode_b)

    return len(antinodes)


def part2(input: str) -> int:
    frequencies = Frequencies(input)
    antinodes = set()

    for frequency in frequencies.groups():
        for a, b in combinations(frequency, 2):
            antinode = a
            d = b - a

            while frequencies.in_range(antinode):
                antinodes.add(antinode)
                antinode = antinode - d

            antinode = a + d

            while frequencies.in_range(antinode):
                antinodes.add(antinode)
                antinode = antinode + d

    return len(antinodes)


TEST_INPUT = """............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............
"""

PART1_TESTS = [
    (TEST_INPUT, 14),
]

PART2_TESTS = [
    (TEST_INPUT, 34),
]
